### ISSUE TITLE
Ranger21 causing loss to spike and model never converges

### ISSUE THREAD

**Issue by aliencaocao** [kozistr/pytorch_optimizer#169]
It is very weird to me but Ranger21 is causing my model (Wav2Vec2 based transformer) to have spikes in loss and it never converges. While some other optimizers from this repo works, like RAdam. I wonder if you know any reason for this or anything different I should have done to use the Ranger21 optimizer properly.

Currently using it with a constant LR of 1e-4 (but ranger21 internally it does warm up and down). Batch size of 8 but with gradient accumulation of 4 so making effective batch size 32, epoch is 100 but at 50 or even 5 same thing happens. I am training with the feature extractor frozen so only the logits layer

Can I also check that the `num_iterations` parameter of Ranger21 refers to number of optimization steps, not epoch? I am currently calculating the steps using this formula: `epoch * len(dataset) / gradient_accmulation steps`

Below is a comparison between Ranger21 and AdamW. Orange is AdamW. wer is word error rate %, ranger21 made the model to reach 100% wer which means the model got everything wrong. Towards the end, the loss of ranger21 seems to be stabilizing but it is actually stuck in a pretty high value and slowly increasing. The graph is not obvious because of the spikes making the scale very small.

![image](https://github.com/kozistr/pytorch_optimizer/assets/20109683/bb46b6a2-223e-4339-a01c-d0f606827268)

**Comment by kozistr** [comment]
Well, there're several reasons that Ranger21 doesn't work (e.g. architecture, training recipes, etc), I guess your training recipe looks fine for me in general. One naive assumption is maybe due to the learning rate or other settings, the gradient goes to near 0 so there's no substantial update. How about using more high lr I guess?

yes. `num_iterations` refers to the total number of optimization steps, `epoch * len(dataset) / gradient_accmulation steps` is correct.

Possibly, there's a bug in the implementation, so, I'll also double-check the Ranger21 implementation.

Thanks for reporting :)

**Comment by aliencaocao** [comment]
I am using fp16 to train, could that be a reason?

**Comment by aliencaocao** [comment]
I think your implementation is correct. I was misled into thinking that the initial lr i used was 1e-4 but it was in fact 1e-3. Reducing that to 1e-4 made it converge well. Thank you and sorry for the false report.

**Comment by kozistr** [comment]
> I think your implementation is correct. I was misled into thinking that the initial lr i used was 1e-4 but it was in fact 1e-3. Reducing that to 1e-4 made it converge well. Thank you and sorry for the false report.

glad you solve the problem! I also found a minor bug in [Ranger21 weight decay](https://github.com/kozistr/pytorch_optimizer/blob/main/pytorch_optimizer/optimizer/ranger21.py#L254) (should be `group['lr']` to `lr`) while double-checking the code, thanks to your report.

**Comment by aliencaocao** [comment]
Nice, mind committing this small fix so I can pull and update my library? Thanks!

**Comment by kozistr** [comment]
> Nice, mind committing this small fix so I can pull and update my library? Thanks!

sure! I'll release a v2.9.1 with the fix soon. I'll let you know when ready.

**Comment by kozistr** [comment]
@aliencaocao hi! I just released [v2.9.1](https://github.com/kozistr/pytorch_optimizer/releases/tag/v2.9.1). you can use it from now :)

Thank you

**Comment by aliencaocao** [comment]
thanks!
